test/tbb/test_collaborative_call_once.cpp

```diff
@@ -0,0 +1,372 @@
+    }
+}
+
+//! Stress test for functor to be called only once
+//! \brief \ref interface \ref requirement \ref stress
+TEST_CASE("only calls once - stress test") {
+    {
+        increment_functor f;
+
+        tbb::collaborative_once_flag flag;
+        std::vector<std::thread> threads;
+        for (std::size_t i = 0; i < utils::get_platform_max_threads()*5; ++i)
+        {
+            threads.push_back(std::thread([&]() {
+                tbb::collaborative_call_once(flag, f);
+            }));
+        }
+        for (auto& thread : threads) {
+            thread.join();
+        }
+        REQUIRE(f.ct == 1);
+    }
+    {
+        increment_functor f;
+
+        utils::SpinBarrier barrier(utils::get_platform_max_threads()*5);
+        tbb::collaborative_once_flag flag;
+        utils::NativeParallelFor(utils::get_platform_max_threads()*5, [&](std::size_t) {
+            for (int i = 0; i < 100; ++i) {
+                REQUIRE(f.ct == i);
+                barrier.wait([&] {
+                    flag.~collaborative_once_flag();
+                    new (&flag) tbb::collaborative_once_flag{};
+                });
+                tbb::collaborative_call_once(flag, f);
+            }
+        });
+    }
+}
+
+#if TBB_USE_EXCEPTIONS
+
+//! Test for collaborative_call_once exception handling
+//! \brief \ref error_guessing
+TEST_CASE("handles exceptions - state reset") {
+
+    std::atomic<bool> b{ false };
+    auto setB = [&b]() { b = true; };
+    auto setBAndCancel = [&b]() {
+        b = true;
+        throw call_once_exception{};
+    };
+
+    tbb::collaborative_once_flag flag;
+    REQUIRE_THROWS_AS(tbb::collaborative_call_once(flag, setBAndCancel), call_once_exception);
+    REQUIRE(b);
+
+    b = false;
+    REQUIRE_THROWS_AS(tbb::collaborative_call_once(flag, setBAndCancel), call_once_exception);
+    REQUIRE(b);
+
+    b = false;
+    tbb::collaborative_call_once(flag, setB);
+    REQUIRE(b);
+
+    b = false;
+    tbb::collaborative_call_once(flag, setB); // Now the call_once flag should be set.
+    REQUIRE(!b);
+
+    b = false;
+    REQUIRE_NOTHROW(tbb::collaborative_call_once(flag, setBAndCancel)); // Flag still set, so it shouldn't be called.
+    REQUIRE(!b);
+}
+
+//! Stress test for collaborative_call_once exception handling
+//! \brief \ref error_guessing \ref stress
+TEST_CASE("handles exceptions - stress test") {
+    std::atomic<int> data{0};
+
+    bool run_again{true};
+
+    auto throwing_func = [&] {
+        utils::doDummyWork(10000);
+        if (data < 100) {
+            data++;
+            run_again = true;
+            throw call_once_exception{};
+        }
+        run_again = false;
+    };
+
+    tbb::collaborative_once_flag flag;
+
+    utils::NativeParallelFor(utils::get_platform_max_threads()*5, [&](std::size_t) {
+        while(run_again) {
+            try {
+                tbb::collaborative_call_once(flag, throwing_func);
+            } catch(const call_once_exception&) {
+            }
```

Comment:
Actually, we are expecting the exception here. In this test case we are trying to stress test algorithm's exception handling, because on exception `collaborative_once_flag` sets its state to `uninitialized`.

Reply:
I suppose we expect `const call_once_exception&` not an arbitrary type (`...`). The idea to enforce that we receive only `const call_once_exception&` (the comment is put on `}` not on the catch - it is supposed to remain as is)